lib/format_parser.rb

```diff
@@ -78,16 +93,34 @@ def self.parse(io, natures: @natures.to_a, formats: @formats.to_a, results: :fir
     end.reject(&:nil?).take(amount)
 
     return results.first if amount == 1
-    # Convert the results from a lazy enumerator to an array.
+    # Convert the results from a lazy enumerator to an Array.
     results.to_a
   end
 
-  def self.parsers_for(natures, formats)
-    # returns lazy enumerator for only computing the minimum amount of work (see :returns keyword argument)
-    @parsers.map(&:new).select do |parser|
-      # Do a given parser contain any nature and/or format asked by the user?
-      (natures & parser.natures).size > 0 && (formats & parser.formats).size > 0
-    end.lazy
+  def self.parsers_for(desired_natures, desired_formats)
+    assemble_parser_set = ->(hash_of_sets, keys_of_interest) {
+      hash_of_sets.values_at(*keys_of_interest).compact.inject(&:+) || Set.new
+    }
+
+    fitting_by_natures = assemble_parser_set[@natures_per_parser, desired_natures]
+    fitting_by_formats = assemble_parser_set[@formats_per_parser, desired_formats]
+    factories = fitting_by_natures & fitting_by_formats
+
+    if factories.empty?
+      raise ArgumentError, "No parsers provide both natures #{desired_natures.inspect} and formats #{desired_formats.inspect}"
+    end
+
+    factories.map { |callable_or_class| instantiate_parser(callable_or_class) }
+  end
+
+  def self.instantiate_parser(callable_or_responding_to_new)
```

Comment:
`callable_or_responding_to_new` => `parser`? I guess it's a more readable name.

🙃

Reply:
Actually here no - I'm kinda hinting at the expected type of the argument. If we say `parser` it is not clear that what we want is either the parser callable or the parser factory.

Reply:
Maybe it's just a cosmetic issue to me as this method can be halved in length, and for that dimensions, maybe the hint is not needed at all. 

😉 

```ruby
def self.instantiate_parser(parser)
  return parser if parser.responds_to?(:call)
  return parser.new if parser.is_a?(Class) && parser.method_defined?(:call)
  raise ArgumentError, '...'
end
```

Reply:
Also, it could mean that a code comment is needed hehehe.

Reply:
Or a monad 😆